openwisp_monitoring/device/base/models.py

```diff
@@ -454,6 +454,20 @@ def handle_disabled_organization(cls, organization_id):
             status='unknown'
         )
 
+    @classmethod
+    def handle_critical_metric(cls, metric):
+        """
+        Updates the device status to 'unknown' if the given metric
+        is a critical metric from the CRITICAL_DEVICE_METRICS setting.
+        """
+        DeviceData = load_model('device_monitoring', 'DeviceData')
+
+        if metric.name in [m['key'] for m in app_settings.CRITICAL_DEVICE_METRICS]:
```

Comment:
There is a property  `related_metrics` of DeviceMonitoring model which returns the related metrics of the device. We shall loop over that queryset and update critical metrics accordingly. 

https://github.com/openwisp/openwisp-monitoring/blob/998c81b4a6e7600d418db2fb96e6a975f72b45cc/openwisp_monitoring/device/base/models.py#L386-L393

Reply:
I would just do the following to get the critical metrics. 

```python 
qs = cls.related_metrics.filter(key__in=app_settings.CRITICAL_DEVICE_METRICS)
```

Reply:
I gave a terrible suggestion earlier. I had the wrong impression that we have a device object here because we are in the AbstractDeviceModel. 

Instead, we have a metric object here. I think, for the sake of readability we should pass `Check` object to the receiver here and then retrieve the related DeviceMonitoring and Metric object in the receiver. 

The `Check` model has a GenericForeignKey relation which includes the object_id of the device. We can get the DeviceMonitoring object like so  

```python
device_monitoring = cls.objects.get(device_id=check.object_id)
device_monitoring.update_status('unknown')
```

I think the this conditional suffices this requirement of the issue, so we don't need to perform another database query 

> Disabling a check for any metric which is listed in OPENWISP_MONITORING_CRITICAL_DEVICE_METRICS (by defaults lists only ping), should put the device in unkown status.

```
if metric.name in [m['key'] for m in app_settings.CRITICAL_DEVICE_METRICS]:
```

Reply:
``
    def connect_check_signals(self):
        from swapper import load_model

        Check = load_model('check', 'Check')
        DeviceMonitoring = load_model('device_monitoring', 'DeviceMonitoring')

        @receiver(post_save, sender=Check)
        def check_post_save_receiver(sender, instance, **kwargs):
            DeviceMonitoring.handle_critical_metric(instance)  

        @receiver(post_delete, sender=Check)
        def check_post_delete_receiver(sender, instance, **kwargs):
            DeviceMonitoring.handle_critical_metric(instance)  
 ``
 
 `` 
     @classmethod
    def handle_critical_metric(cls, check):
        """
        Updates the device status to 'unknown' if the given check
        is for a critical metric from the CRITICAL_DEVICE_METRICS setting
        and the check is not active.
        """
        from django.conf import settings

        critical_metrics = getattr(
            settings,
            'OPENWISP_MONITORING_CRITICAL_DEVICE_METRICS',
            [{'key': 'ping', 'field_name': 'reachable'}],
        )
        # Check if the check is associated with a critical metric
        if check.metric in [m['key'] for m in critical_metrics] and not check.is_active:
            try:
                device_monitoring = cls.objects.get(device_id=check.object_id)
                device_monitoring.update_status('unknown')
            except cls.DoesNotExist:
                pass
 ``